File: ICPC_Mirrors/Nitc_9.0/nwerc2019all/firetrucksarered/submissions/accepted/Paul.java
import java.util.*;
import java.util.stream.*;

public class Paul {
	static class UnionFind {
		int[] parent, size;

		public UnionFind(int n) {
			parent = IntStream.range(0,n).toArray();
			size = IntStream.range(0,n).map(i -> 1).toArray();
		}

		int find(int a) {
			if (a == parent[a]) return a;
			return parent[a] = find(parent[a]);
		}

		boolean union(int a, int b) {
			a = find(a);
			b = find(b);
			if (a == b) return false;
			if (size[a] > size[b]) {
				parent[b] = a;
				size[a] += size[b];
			} else {
				parent[a] = b;
				size[b] += size[a];
			}
			return true;
		}
	}

	static class Edge {
		int a, b, num;

		public Edge(int a, int b, int num) {
			this.a = a;
			this.b = b;
			this.num = num;
		}
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		
		Map<Integer,List<Integer>> occurrences = new HashMap<>();
		UnionFind uf = new UnionFind(n+1);
		List<Edge> result = new ArrayList<>();

		for (int a = 1; a <= n; a++) {
			int m = sc.nextInt();
			for (int j = 0; j < m; j++) {
				int x = sc.nextInt();
				if (!occurrences.containsKey(x)) {
					occurrences.put(x, new ArrayList<>());
				}
				occurrences.get(x).add(a);
			}
		}
		
		for (var entry: occurrences.entrySet()) {
			int a = entry.getValue().get(0);
			for (int b: entry.getValue()) {
				if (uf.union(a,b)) {
					result.add(new Edge(a,b,entry.getKey()));
				}
			}
		}

		if (uf.size[uf.find(1)] < n) {
			System.out.println("impossible");
		} else {
			for (Edge e: result) {
				System.out.println(e.a + " " + e.b + " " + e.num);
			}
		}
	}
}

File: ICPC_Mirrors/Nitc_9.0/nwerc2019all/firetrucksarered/submissions/accepted/paul.cpp
#include <bits/stdc++.h>
using namespace std;

struct union_find {
	vector<int> parent, size;

	union_find(int n): parent(n), size(n,1) {
		iota(begin(parent), end(parent), 0);
	}


	int find(int a) {
		if (a == parent[a]) return a;
		return parent[a] = find(parent[a]);
	}

	bool unite(int a, int b) {
		a = find(a), b = find(b);
		if (a == b) return false;
		if (size[a] > size[b]) swap(a,b);
		parent[a] = b;
		size[b] += size[a];
		return true;
	}
};

int main() {
	int n; cin >> n;
	
	map<int,vector<int>> occurrences;
	union_find uf(n+1);
	vector<tuple<int,int,int>> result;

	for (int a = 1; a <= n; a++) {
		int m; cin >> m;
		for (int j = 0; j < m; j++) {
			int x; cin >> x;
			occurrences[x].push_back(a);
		}
	}
	
	for (auto [x,as]: occurrences) {
		int a = as[0];
		for (int b: as) if (uf.unite(a,b)) {
			result.emplace_back(a,b,x);
		}
	}

	if (uf.size[uf.find(1)] < n) {
		cout << "impossible" << endl;
	} else {
		for (auto [a,b,x]: result) {
			cout << a << " " << b << " " << x << endl;
		}
	}
}
